3p-ecosystem/src/main/java/com/google/homesampleapp/screens/device/DeviceFragment.kt

```diff
@@ -332,68 +335,206 @@ class DeviceFragment : Fragment() {
   }
 
   // -----------------------------------------------------------------------------------------------
-  // UI update functions
+  // Composables
+
+  @Composable
+  //private fun DeviceRoute(selectedDeviceViewModel: SelectedDeviceViewModel) {
+  private fun DeviceRoute() {
+    // Observes values needed by the DeviceScreen.
+    val selectedDeviceId by selectedDeviceViewModel.selectedDeviceIdLiveData.observeAsState()
+    val selectedDevice by selectedDeviceViewModel.selectedDeviceLiveData.observeAsState()
+    val lastUpdatedDeviceState by devicesStateRepository.lastUpdatedDeviceState.observeAsState()
+    val shareDeviceStatus by viewModel.shareDeviceStatus.observeAsState()
+
+    binding.topAppBar.title = selectedDevice?.device?.name
 
-  private fun updateDeviceInfo(deviceState: DeviceState?) {
+    DeviceScreen(selectedDeviceId, selectedDevice, lastUpdatedDeviceState, shareDeviceStatus)
+  }
+
+  @Composable
+  private fun DeviceScreen(
+    selectedDeviceId: Long?,
+    deviceUiModel: DeviceUiModel?,
+    deviceState: DeviceState?,
+    shareDeviceStatus: TaskStatus?
+  ) {
     // The DeviceUiModel is not updated whenever we observe changes in the state of the device.
     // This is an issue for the "Inspect Device" onClick listener which relies on the device
     // state to decide whether to show a dialog stating that the device is offline and therefore
     // the inspect screen cannot be shown, or go show the inspect information (when device is
     // online).
     // This is why the state of the device is cached in local variables.
-    if (selectedDeviceViewModel.selectedDeviceIdLiveData.value == -1L) {
+    if (selectedDeviceId == null || selectedDeviceId == -1L) {
```

Comment:
The local variables `isOnline` and `isOn` can be brought into the composition and observed by using:

```
var isOnline by remember { mutableStateOf(false) }
var isOn by remember { mutableStateOf(false) }
```

The benefit of doing this is efficiency. When held in class fields that aren't backed by Compose state, the Compose compiler can't guarantee when they've changed, so it eagerly recomposes often to make sure it shows the most up to date state.

By moving them into compose state backed variables in the composition, we can avoid this. To make sure its updated when either the deviceUiModel or device state changes, `LaunchedEffect` can be used:

```
LaunchedEffect(selectedDeviceId, deviceUiModel, deviceState) {
    if (selectedDeviceId == null || selectedDeviceId == -1L) {
        isOnline = false
        return@LaunchedEffect
        
        deviceUiModel?.let {
            isOnline = when (deviceState) {
                  null -> deviceUiModel.isOnline
                  else -> deviceState.online
            }
            isOn =  when (deviceState) {
                null -> deviceUiModel.isOn
                else -> deviceState.on
            }
    }
}
```

This also moves the early return into the lambda of the `LaunchedEffect` instead of in the Composition.

Reply:
Thanks for the suggestion and for the clear explanation. Fixed in the latest commit.